### PR TITLE
[ftxui] Add new port (fix #16327)

### LINKED ISSUE #16327: [New Port Request] <FTXUI>
Library name: FTXUI

Library description: Functional Terminal (X) User interface. A simple C++ library for terminal based user interface. 

Source repository URL: https://github.com/ArthurSonzogni/FTXUI

Project homepage (if different from the source repository):

Anything else that is useful to know when adding (such as optional features the library may have that should be included):

### PR DESCRIPTION
**Pull request by myd7349**
**Describe the pull request**

- #### What does your PR fix?  
  Fixes #16327

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?  
  <all / linux, windows, ...>, <Yes/No>

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?  
  `Your answer`

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?  
  <Yes / I am still working on this PR>

**If you are still working on the PR, open it as a Draft: https://github.blog/2019-02-14-introducing-draft-pull-requests/**

### CODE REVIEW

File: ports/ftxui/vcpkg.json [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+  "dependencies": [
+    {
+      "name": "vcpkg-cmake",
+      "host": true
+    },
+    {
+      "name": "vcpkg-cmake-config",
+      "host": true
+    }
+  ]
```
Comment (ArthurSonzogni) [inline ports/ftxui/vcpkg.json:16]: Out of curiosity, what are those about?

**Review by ArthurSonzogni — COMMENTED** [review]
Thanks for working on this!
I wanted to start doing the same, but since the work has already been done here, better doing the same.

I see this uses version-date.
At some point, I would like to be able to adhere to semantic versioning, when I will have reached 1.0.
Do you think, this will still be possible later to switch to it?

File: ports/ftxui/vcpkg.json (hunk 1 above)
Comment (JonLiu1993) [reply ports/ftxui/vcpkg.json:16]: When using vcpkg_cmake_config_fixup(), you need to add vcpkg-cmake-config, and use other new functions only need to add vcpkg-cmake

**Comment by myd7349** [comment]
Hi! @ArthurSonzogni Thanks for your awesome work.
Yeah. I'd like to switch to v1.0 if it was released in the near future.

File: ports/ftxui/vcpkg.json [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+  "name": "ftxui",
+  "version-date": "2021-06-09",
+  "description": "C++ Functional Terminal User Interface",
+  "homepage": "https://github.com/ArthurSonzogni/FTXUI",
+  "license": "MIT",
+  "dependencies": [
+    {
+      "name": "vcpkg-cmake",
+      "host": true
+    },
```
Comment (MichaelGoulding) [inline ports/ftxui/vcpkg.json:11]:
```suggestion
```

Comment (MichaelGoulding) [reply ports/ftxui/vcpkg.json:11]: I had to remove this to get it to work for me.

**Comment by JonLiu1993** [comment]
@myd7349 ,Thanks for your pr.
Could you take a look?
[install-x64-uwp-dbg-out.log](https://github.com/microsoft/vcpkg/files/6659703/install-x64-uwp-dbg-out.log)

**Comment by myd7349** [comment]
> 
> 
> @myd7349 ,Thanks for your pr.
> Could you take a look?
> [install-x64-uwp-dbg-out.log](https://github.com/microsoft/vcpkg/files/6659703/install-x64-uwp-dbg-out.log)

Hi! @JonLiu1993 Thanks for your review.

According to the build log:
> D:\buildtrees\ftxui\src\719840ffdc-7f05139eb7.clean\src\ftxui\screen\terminal.cpp(56,37): error C2039: 'getenv': is not a member of 'std' [D:\buildtrees\ftxui\x64-uwp-dbg\screen.vcxproj]
D:\buildtrees\ftxui\src\719840ffdc-7f05139eb7.clean\src\ftxui\screen\terminal.cpp(60,32): error C2039: 'getenv': is not a member of 'std' [D:\buildtrees\ftxui\x64-uwp-dbg\screen.vcxproj]

ftxui can not be built successfully on UWP, since `getenv` is unavailable on UWP:
https://docs.microsoft.com/en-us/cpp/cppcx/crt-functions-not-supported-in-universal-windows-platform-apps?view=msvc-160

**Comment by myd7349** [comment]
Hi! @MichaelGoulding That was strange. Could you verify your `vcpkg-tool` version?
> ./vcpkg version

File: ports/ftxui/portfile.cmake [hunk 3]
```diff
@@ -0,0 +1,26 @@
+vcpkg_check_linkage(ONLY_STATIC_LIBRARY)
+
```
Comment (JonLiu1993) [inline ports/ftxui/portfile.cmake:2]:
```suggestion
vcpkg_fail_port_install(ON_TARGET "uwp") 
```

File: versions/f-/ftxui.json [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+    {
+      "git-tree": "6aa3d0f687a5cd121540e8e2bc158d0aaf77e905",
+      "version-date": "2021-06-16",
+      "port-version": 0
+    },
+    {
+      "git-tree": "4f720eb108815befe381654803dbdae18ee0615d",
+      "version-date": "2021-06-09",
+      "port-version": 0
+    }
```
Comment (JonLiu1993) [inline versions/f-/ftxui.json:12]: Please remove this

**Review by JonLiu1993 — CHANGES_REQUESTED** [review]

**Comment by myd7349** [comment]
Hi! @JonLiu1993 Thanks for your review.

**Review by JonLiu1993 — APPROVED** [review]

**Comment by myd7349** [comment]
One more note:
https://github.com/microsoft/vcpkg/issues/16327#issuecomment-863945954
https://github.com/microsoft/vcpkg/issues/16327#issuecomment-865455700

File: ports/ftxui/vcpkg.json [hunk 5]
```diff
@@ -0,0 +1,18 @@
+{
+  "name": "ftxui",
+  "version-date": "2021-06-17",
+  "description": "C++ Functional Terminal User Interface",
+  "homepage": "https://github.com/ArthurSonzogni/FTXUI",
+  "license": "MIT",
+  "supports": "!uwp",
```
Comment (JackBoosY) [inline ports/ftxui/vcpkg.json:7]: Is the upstream doesn't support uwp or it doesn't support uwp in vcpkg currently?

File: ports/ftxui/portfile.cmake [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+        -DFTXUI_BUILD_DOCS=OFF
+)
+
+vcpkg_cmake_install()
+
+vcpkg_cmake_config_fixup(CONFIG_PATH lib/cmake/${PORT})
+
+file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/include" "${CURRENT_PACKAGES_DIR}/debug/share")
+
+file(INSTALL ${SOURCE_PATH}/LICENSE DESTINATION ${CURRENT_PACKAGES_DIR}/share/${PORT} RENAME copyright)
```
Comment (JackBoosY) [inline ports/ftxui/portfile.cmake:28]:
```suggestion
file(INSTALL "${SOURCE_PATH}/LICENSE" DESTINATION "${CURRENT_PACKAGES_DIR}/share/${PORT}" RENAME copyright)

```

**Review by JackBoosY — CHANGES_REQUESTED** [review]

**Comment by myd7349** [comment]
Hi! @*JackBoosY*

According to the build log:
>
D:\buildtrees\ftxui\src\719840ffdc-7f05139eb7.clean\src\ftxui\screen\terminal.cpp(56,37):
error C2039: 'getenv': is not a member of 'std'
[D:\buildtrees\ftxui\x64-uwp-dbg\screen.vcxproj]
D:\buildtrees\ftxui\src\719840ffdc-7f05139eb7.clean\src\ftxui\screen\terminal.cpp(60,32):
error C2039: 'getenv': is not a member of 'std'
[D:\buildtrees\ftxui\x64-uwp-dbg\screen.vcxproj]

ftxui can not be built successfully on UWP, since getenv is unavailable on
UWP:
https://docs.microsoft.com/en-us/cpp/cppcx/crt-functions-not-supported-in-universal-windows-platform-apps?view=msvc-160

Jack·Boos·Yu ***@***.***> 于2021年6月22日周二 下午3:26写道：

**Comment by myd7349** [comment]
Hi! @JackBoosY Thanks. Updated.

File: ports/ftxui/vcpkg.json (hunk 2 above)
Comment (JackBoosY) [reply ports/ftxui/vcpkg.json:12]: It always be the host dependency.

**Review by JackBoosY — APPROVED** [review]

**Pull request merged**